### PR TITLE
Test PR

### PR DESCRIPTION
**Pull request by cb1423**

### CODE REVIEW

File: UserService.groovy [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+        } else if (user.currentlyPlaying != playing) {
+            user.lastChanged = new Date()
+            user.status = UserStatus.ACTIVE
+        }
+
+        user.currentlyPlaying = user.currentlyPlaying ?: new CurrentlyPlaying(user: user)
+
+        user.currentlyPlaying.track = recordService.updateTrack(playing?.track)
+        user.currentlyPlaying.progress = playing.progress
+    }
```
Comment (cb1423) [inline UserService.groovy:190]: Seeing how code block comments work in the plugin

File: UserService.groovy [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    User getOrCreateUser(Profile profile, Authentication authentication) {
+        User user
+        if (getUser(profile.id)) {
+            user = getUser(profile.id)
+            setAuthentication(user, authentication)
+        } else {
+            user = createUser(profile)
+            setAuthentication(user, authentication)
+
+            // TODO Only for development, remove before production
```
Comment (ghost) [inline UserService.groovy:50]: Comments == baaaaaaaad

File: UserService.groovy [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+            smartPlaylistService.importFromSpotify(user)
+            smartPlaylistService.deleteAllSmartPlaylists(user)
+            // TODO Only for development, remove before production
```
Comment (ghost) [inline UserService.groovy:53]: Good comment

File: UserService.groovy [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+        return apiService.getProfile(accessToken)
+    }
+
+    void logon(User user) {
+        user.status = UserStatus.ACTIVE
+        user.lastChanged = new Date()
+        user.save(flush: true)
+        userInfoServiceProxy.id = user.id
+        log.info "${user?.displayName} has logged on"
+    }
```
Comment (ghost) [inline UserService.groovy:78]: Tell me more about this method, it's interesting

File: UserService.groovy [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+
+    Authentication getAuthentication(String code, String state) {
+        if (state != this.state && state != null) {
+            throw new AuthenticationException('Invalid state', 'The server did not return the same state that it was supplied with')
+        }
+        log.debug("Authorisation code: $code")
+        log.debug("state received: $state")
+        this.state = null
+        Token token = apiService.getToken(code)
+        return token.authentication
```
Comment (ghost) [inline UserService.groovy:127]: Do dis

```suggestion
        return apiService.getToken(code)?.authentication
```

File: UserService.groovy [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+    }
+
+    void setAuthentication(User user, Authentication authentication) {
+        Authentication auth = Authentication.findOrCreateByUser(user)
+        auth.with {
+            accessToken = authentication.accessToken
+            refreshToken = authentication.refreshToken
+            expiry = authentication.expiry
+        }
```
Comment (ghost) [inline UserService.groovy:136]: Adding this comment as part of a review (hint: "Pending" means no-one else can see it!)

File: UserService.groovy [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+            return
+        } else if (user.currentlyPlaying?.isReset(playing)) {
+            user.lastRecorded?.delete()
+            user.lastRecorded = null
+        } else if (user.currentlyPlaying != playing) {
+            user.lastChanged = new Date()
+            user.status = UserStatus.ACTIVE
+        }
+
+        user.currentlyPlaying = user.currentlyPlaying ?: new CurrentlyPlaying(user: user)
```
Comment (ghost) [inline UserService.groovy:186]:
```suggestion
        user.currentlyPlaying = (user.currentlyPlaying ?: new CurrentlyPlaying(user: user))
```

File: UserService.groovy [hunk 8]
```diff
@@ -0,0 +1,236 @@
+package smartify
```
Comment (ghost) [inline UserService.groovy:1]: wow. much smart :dog:

**Review by ghost — CHANGES_REQUESTED** [review]
Best code I've ever seen!

Not good enough, requesting changes

**Pull request closed without merging**